--- /usr/local/lib/perl5/5.8.0/mach/B/C.pm	Sat Oct 19 10:17:53 2002
+++ C.pm	Sat Oct 19 10:17:58 2002
@@ -1619,6 +1619,15 @@
     }
 EOT
     }
+    else {
+	print <<EOT;
+    if ((tmpgv = gv_fetchpv("0",TRUE, SVt_PV))) {/* $0 */
+        tmpsv = GvSV(tmpgv);
+        sv_setpv(tmpsv, argv[0]);
+        SvSETMAGIC(tmpsv);
+    }
+EOT
+    }
 
     print <<'EOT';
     if ((tmpgv = gv_fetchpv("\030",TRUE, SVt_PV))) {/* $^X */
